pkg/registry/apis/query/runner/direct.go

```diff
@@ -0,0 +1,222 @@
+	return &directRunner{
+		pluginClient: pluginClient,
+		pCtxProvider: pCtxProvider,
+	}
+}
+
+func NewDirectRegistry(pluginStore pluginstore.Store,
+	dataSourcesService datasources.DataSourceService,
+) DataSourceRegistry {
+	return &directRegistry{
+		pluginStore:        pluginStore,
+		dataSourcesService: dataSourcesService,
+	}
+}
+
+// ExecuteQueryData implements QueryHelper.
+func (d *directRunner) ExecuteQueryData(ctx context.Context,
+	// The k8s group for the datasource (pluginId)
+	datasource schema.GroupVersion,
+
+	// The datasource name/uid
+	name string,
+
+	// The raw backend query objects
+	query []v0alpha1.GenericDataQuery,
+) (*backend.QueryDataResponse, error) {
+	queries, dsRef, err := legacydata.ToDataSourceQueries(v0alpha1.GenericQueryRequest{
+		Queries: query,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if dsRef != nil && dsRef.UID != name {
+		return nil, fmt.Errorf("expected query body datasource and request to match")
+	}
+
+	// NOTE: this depends on uid unique across datasources
+	settings, err := d.pCtxProvider.GetDataSourceInstanceSettings(ctx, name)
+	if err != nil {
+		return nil, err
+	}
+
+	pCtx, err := d.pCtxProvider.PluginContextForDataSource(ctx, settings)
+	if err != nil {
+		return nil, err
+	}
+
+	return d.pluginClient.QueryData(ctx, &backend.QueryDataRequest{
+		PluginContext: pCtx,
+		Queries:       queries,
+	})
+}
+
+// GetDatasourceAPI implements DataSourceRegistry.
+func (d *directRegistry) GetDatasourceAPI(pluginId string) (schema.GroupVersion, error) {
+	d.pluginsMu.Lock()
+	defer d.pluginsMu.Unlock()
+
+	if d.plugins == nil {
+		err := d.updatePlugins()
+		if err != nil {
+			return schema.GroupVersion{}, err
+		}
+	}
+
+	var err error
+	gv, ok := d.apis[pluginId]
+	if !ok {
+		err = fmt.Errorf("no API found for id: " + pluginId)
+	}
+	return gv, err
+}
+
+// GetDataSources implements QueryHelper.
+func (d *directRegistry) GetDataSources(ctx context.Context, namespace string, options *internalversion.ListOptions) (*v0alpha1.DataSourceList, error) {
+	info, err := request.NamespaceInfoFrom(ctx, true)
+	if err != nil {
+		return nil, err
+	}
+	datasources, err := d.dataSourcesService.GetDataSources(ctx, &datasources.GetDataSourcesQuery{
+		OrgID: info.OrgID,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	results := &v0alpha1.DataSourceList{
+		ListMeta: metav1.ListMeta{
+			ResourceVersion: fmt.Sprintf("%d", time.Now().UnixMilli()),
+		},
+	}
+	for _, real := range datasources {
+		gv := d.apis[real.Type]
+
+		ds := v0alpha1.DataSource{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:              real.UID,
+				CreationTimestamp: metav1.NewTime(real.Created),
+			},
+			Title: real.Name,
+			Group: gv.Group,
+			// Health: &v0alpha1.HealthCheck{
+			// 	Status:  "OK",
+			// 	Checked: time.Now().UnixMilli(),
+			// },
+		}
+
+		results.Items = append(results.Items, ds)
+	}
+
+	return results, nil
+}
+
+// GetDatasourcePlugins no namespace? everything that is available
```

Comment:
It's a bit unclear what you're trying to achieve, any datasource api server available or datasource backend plugin installed per tenant?

Reply:
In the "direct" mode, this seems pretty silly -- i mean, just use the wire services!

However -- the value of this approach is that the api-server is not able to decide how to resolve each query.  It may be local/multi-tenant or delegate to a single tenant grafana instance.

Reply:
I renamed this to `GetDatasourceApiServers` -- hopefully more clear